README.md
# Smart Industry: analysis of anomaly detection simulation
This repository contains the code and results of simulating live anomaly detection performed on the [AURSAD](https://zenodo.org/record/4487073) dataset.
To view and experiment with different settings of the simulation, an interactive web app is provided.

The project emulates the workings of a system for x number of cycles, where every cycle consists of reading a new sample
to the buffer, making a prediction and then appending the results to be summarily saved at the end of the simulation.
The simulation measures the accuracy of the predictions, as well as the run times for each cycle.
The experiments are performed for the binary version of the dataset, which means that the operation can be normal or 
anomalous, without distinction between the different types of anomalies.
The data is sampled at 50 Hz.

Based on these results, the web app allows experimenting with applying different methods to create the optimal system response accuracy.

## Requirements
The solution has been intended for and tested on Ubuntu & Windows 10. 
For training new models, a Nvidia GPU with installed and 
properly configured cuDNN drivers is recommended. 
More on how to install those drivers can be found [here](https://docs.nvidia.com/deeplearning/cudnn/install-guide/index.html).

## Installation instructions
1. Install Anaconda from [here](https://www.anaconda.com/products/individual)
2. Clone the Git repository to chosen directory
3. Open the Anaconda prompt and navigate to the project's directory
4. Enter the following to create and activate the new virtual environment:
```bash
> conda env create -f environment.yml
> conda activate anomaly_simulation
```
5. Download the [AURSAD](https://zenodo.org/record/4487073) dataset and move it to the directory anomaly_simulation/Data
6. All the following commands must be run from the activated virtual environment

## How to use it?
The project can be used in several ways. 

### Analysing the included simulation results
The easiest and quickest way is to run the web app and explore the included simulation results. 
To do this, in your Python enabled terminal (terminal or IDE terminal on Ubuntu, Anaconda Prompt or IDE terminal on Windows) 
run the following command from the project's directory:
```bash
/anomaly_simulation/Source> python visualisation/app.py
```
The results of this command should look similar to this:
```bash
Dash is running on http://127.0.0.1:8050/

 * Serving Flask app "app" (lazy loading)
 * Environment: production
   WARNING: This is a development server. Do not use it in a production deployment.
   Use a production WSGI server instead.
 * Debug mode: off
 * Running on http://127.0.0.1:8050/ (Press CTRL+C to quit)
```
Copy the IP address to your web browser. The visualisation may take a minute to load and process the data. It should look like this:
![alt text][web_app]

[web_app]: web_app.png 

In the top left section model parameters can be chosen:
* Model window size - describes how many previous samples the model takes into account to predict the class of the next sample
* Model dimensionality - describes how many dimensions the input data has. Original data in the AURSAD dataset contains 125 dimensions, any number
below that is achieved by applying PCA dimensionality reduction method.
* Screwdriver data only - only the screwdriver's sensors data will be used
* Binary data - binary labels will be used (normal and anomaly)
  
The app then proceeds to find the best model with such parameters from those available under the anomaly_simulation/Zoo/Results/runs directory.
  
The dropdown box and slider below the model controls provide the means to manipulate the parameters of the sliding window averages 
with the aim of creating the most accurate system response.
The threshold box changes the decision threshold value.
The system calculates the mean label value in each window, and then applies that threshold to it.
If the value exceed the threshold, the sample on which the window is centered is assigned an anomaly label, otherwise 
it gets a normal label.

The figures are described as follows:
* The four figures below show the results of applying these methods on the raw model predictions output.
* The top left figure presents the labels predicted by the system compared to the true labels. 0 means a normal operation, 1 indicates an anomaly.
* The top right figure shows the Cumulative Mean Average of system response accuracy. It allows the user to see how certain trends in data
  affect the model's accuracy. The last value in this graph corresponds to the 'simulation accuracy' value displayed in the box above it, 
  and represents the average accuracy for the whole simulation.
* The bottom left figure compares the results of the raw model output, and the processed system response. The figure on the 
  left is a benchmark upon which we want to improve using the post-processing.
* The bottom right figure presents run times for each cycle.

### Performing new simulations on the included, pre-trained models
To run a new simulation using one of the available models, run the following in your terminal
```bash
/anomaly_simulation/Source> python simulation.py --model_dir model_name
```
The simulation requires the user to give the starting symbols of the mo/  \el to use in the simulation, and allows to 
set several other parameters. 
```bash
/anomaly_simulation/Source>python simulation.py --help
Usage: simulation.py [OPTIONS]

  Starts the processes and runs the simulations

Options:
  --version                   Show the version and exit.
  --cycle_count INTEGER       Number of simulation cycles. Default: 50000
  --binary_labels BOOLEAN     True for binary labels, False for multi-class.
                              Default: True

  --window INTEGER            Rolling window size. Default: 500
  --n_dim INTEGER             Data dimensionality. Default: 60
  --screwdriver_only BOOLEAN  If True use only the data from screwdriver
                              sensors. Default: False

  --n_cpu INTEGER             Number of cpu cores to use if running multiple
                              simulations: Default: max - 2

  --run_mode TEXT             Determines how many simulations will be run:
                              single will run 1 simulation with a specified
                              model, all will run all models, fill will run
                              all models that do not havesimulation results
                              yet. Default: single

  --model_dir TEXT            The first symbols of folder name in
                              Results/runs. Default: 27760

  --verbose INTEGER           Logging level, from {0 (debug), 1 (info), 2
                              (warning), 3 (error)}. Default: 1

  --help                      Show this message and exit.
```

This will run the simulation and save its results to anomaly_simulation/Results. 
This simulation will be then available for the web app to automatically detect and load.

### Training new models
To create a new variant of an already created model, open the anomaly_simulation/Zoo/guild.yml file.
This is the file where model parameters are defined.
Example ResNet configuration looks like this:
```bash
- model: ResNet
  description: ResNet classifier
  operations:
    train:
      description: Train ResNet on sliding windows
      main: ResNet/train
      flags:
        window:
          description: The sliding window size
          default: 100
        horizon:
          description: Prediction horizon
          default: 1
        dimensionality:
          description: Data dimensionality
          default: 60
        binarize:
          description: Use binary labels
          default: False
        optimizer:
          description: Loss optimizer
          default: 'adam'
        learning_rate:
          description: Initial learning rate
          default: 0.001
        n_feature_maps:
          description: Number of feature maps
          default: 64
        dev:
          description: Fast development run
          default: False
        remote:
          description: If running on remote
          default: False
      output-scalars:
        train_f1_avg: 'Train f1_avg: (\value)'
        test_f1_avg: 'Test f1_avg: (\value)'
```
There are two ways now to run the model with new parameters.
One is to change the default values of the parameters and save the file.
Next, in your terminal run the following to see available models:
```bash
anomaly_simulation/Zoo> guild operations
Refreshing flags...
Pure_LSTM:train  Train Pure_LSTM on sliding windows
ResNet:train     Train ResNet on sliding windows
TABL:train       Train TABL on sliding windows
```

To run the ResNet model type, with <path> being the absolute path of the anomaly_simulation/Zoo/Results folder, into the terminal:
```bash
anomaly_simulation/Zoo> guild -H <path> run ResNet:train
Refreshing flags...
You are about to run ResNet:train
  binarize: no
  dev: no
  dimensionality: 60
  horizon: 1
  learning_rate: 0.001
  n_feature_maps: 64
  optimizer: adam
  remote: no
  window: 100
Continue? (Y/n) # Tyoe Y and press enter to commence model training
```
This will run the model with the default flags as configured in the guild.yml file.

To run a model with grid search of certain attributes, or change the flags without changing the default values, run the previous command with additional options:
```bash
anomaly_simulation/Zoo> guild -H <path> run ResNet:train window=[100,200,500]
You are about to run ResNet:train as a batch (3 trials)
  binarize: no
  dev: no
  dimensionality: 60
  horizon: 1
  learning_rate: 0.001
  n_feature_maps: 64
  optimizer: adam
  remote: no
  window: [100, 200, 500]
Continue? (Y/n) # Tyoe Y and press enter to commence model training
```

This will automatically queue and train 3 separate ResNet models with different window sizes.

**Adding new model construction**
To add a new model that is compatible with the project, please inspect and recreate the pattern of the files associated with the
already implemented models.

---
<sup>For support, create a new Issue on this repository's GitHub page or email bl@ece.au.dk.</sup>

CHANGELOG.md
# Changelog
All notable changes will be documented in this file.

## [Unreleased]

## [0.2.1] - 07/04/2021
### Changed
* Logging method changed from printing to logger

## [0.2.0] - 05/04/2021
### Added
* Batch simulation
    * Multiprocessing
    * Single, fill and all modes

## [0.1.0] - 31/03/2021
### Added
* Changelog file


Source/simulation.py
__doc__ = """
Script responsible for running the entire simulation.
"""

__version__ = """ v0.1.0 First release"""
__version__ = """ v0.2.0 Added multiprocessing and batch simulations"""
__version__ = """ v0.2.1 Added logger"""

import os

# Tensorflow logging level
os.environ["TF_CPP_MIN_LOG_LEVEL"] = "2"

import tensorflow as tf

physical_devices = tf.config.list_physical_devices('GPU')
if len(physical_devices) > 0:
    tf.config.experimental.set_memory_growth(physical_devices[0], enable=True)

import sys
from tqdm.contrib.concurrent import process_map
import tqdm
import click
import tensorflow as tf
import aursad
import multiprocessing as mp
import pandas as pd

from tensorflow.keras.models import load_model
from pathlib import Path, PurePath

from sensor import Sensor
from hapDev import HapDev

sys.path.append('../')
import Zoo.meta as meta
from Zoo.TABL.Layers import BL, TABL
from Source.visualisation.visualisation import plot_simulation_history
from simulation_utils import list_simulations, setup_logging, group_models


def validate_model_dir(ctx, param, value):
    for x in os.listdir('../Zoo/Results/runs'):
        if x.startswith(value):
            value = x
            return value

    raise click.BadParameter('no such model exists')


@click.command()
@click.version_option(version=__version__)
@click.option('--cycle_count', default=50000, type=int, help='Number of simulation cycles. Default: 50000')
@click.option('--binary_labels', default=True, type=bool, help='True for binary labels, False for multi-class. '
                                                               'Default: True')
@click.option('--window', default=500, type=int, help='Rolling window size. Default: 500')
@click.option('--n_dim', default=60, type=int, help='Data dimensionality. Default: 60')
@click.option('--screwdriver_only', default=False, type=bool, help='If True use only the data from screwdriver sensors.'
                                                                   ' Default: False')
@click.option('--n_cpu', default=mp.cpu_count() - 2, type=int, help='Number of cpu cores to use if running'
                                                                    ' multiple simulations: Default: max - 2')
@click.option('--run_mode', default='single', type=str, help='Determines how many simulations will be run: single will '
                                                             'run 1 simulation with a specified model, all will run '
                                                             'all models, fill will run all models that do not have'
                                                             'simulation results yet. Default: single')
@click.option('--model_dir',
              default='27760',
              callback=validate_model_dir,
              type=str,
              help='The first symbols of folder name in Results/runs. Default: 27760')
@click.option('--verbose', default=0, type=int, help='Logging level, from {0 (debug), 1 (info), 2 (warning), 3 (error)}'
                                                     '. Default: 1')
def main(cycle_count, binary_labels, window, n_dim, screwdriver_only, n_cpu, run_mode, model_dir, verbose):
    """
    Starts the processes and runs the simulations
    """
    logger = setup_logging(verbose)
    logger.info('Simulation parameters: \n'
                 f'Run mode: {run_mode} \n'
                 f'Model signature: {model_dir} \n'
                 f'Cycle count: {cycle_count} \n'
                 f'Binary labels: {binary_labels} \n'
                 f'Window size: {window} \n'
                 f'Data dimensionality: {n_dim} \n'
                 f'Screwdriver data only: {screwdriver_only} \n'
                 f'Maximum cores used: {n_cpu} \n')
    click.confirm('Confirm to run the simulation with the above parameters', abort=True)

    simulation_args = {'cycle_count': cycle_count,
                       'binary_labels': binary_labels,
                       'model_dir': model_dir,
                       'window': window,
                       'n_dim': n_dim,
                       'screwdriver_only': screwdriver_only}
    if run_mode == 'single':
        # Whether to reduce dimensionality
        if simulation_args['n_dim'] < 125:
            reduce_dim = True
        else:
            reduce_dim = False

        logger.info('Loading data')
        _, _, test_x, test_y = aursad.get_dataset_numpy(path=Path(meta.data_path),
                                                        reduce_dimensionality=reduce_dim,
                                                        n_dimensions=n_dim,
                                                        subsample_data=True,
                                                        subsample_freq=2,
                                                        pad_data=False,
                                                        normal_samples=1,
                                                        damaged_samples=1,
                                                        assembly_samples=1,
                                                        missing_samples=1,
                                                        damaged_thread_samples=0,
                                                        loosening_samples=0,
                                                        drop_extra_columns=True,
                                                        onehot_labels=False,
                                                        binary_labels=binary_labels,
                                                        screwdriver_only=screwdriver_only)
        logger.info('Loaded data')
        simulation_args['test_x'] = test_x
        simulation_args['test_y'] = test_y
        single_simulation(simulation_args)
    elif run_mode == 'fill':
        model_list = list_simulations(fill_mode=True, cycle_count=cycle_count)
        n_simulations = len(model_list)

        for i, d in enumerate(model_list):
            model_list[i]['cycle_count'] = cycle_count

        # Group models by the data settings - due to RAM constraints only models using identical datasets can be run
        # concurrently
        grouped_models = group_models(model_list)

        logger.info(f'Running {n_simulations} simulations in {len(grouped_models)} groups')

        for k, v in tqdm.tqdm(grouped_models.items(), desc='Simulation groups'):
            # Whether to reduce dimensionality
            if v[0]['n_dim'] < 125:
                reduce_dim = True
            else:
                reduce_dim = False

            _, _, test_x, test_y = aursad.get_dataset_numpy(path=Path(meta.data_path),
                                                            reduce_dimensionality=reduce_dim,
                                                            n_dimensions=v[0]['n_dim'],
                                                            subsample_data=True,
                                                            subsample_freq=2,
                                                            pad_data=False,
                                                            normal_samples=1,
                                                            damaged_samples=1,
                                                            assembly_samples=1,
                                                            missing_samples=1,
                                                            damaged_thread_samples=0,
                                                            loosening_samples=0,
                                                            drop_extra_columns=True,
                                                            onehot_labels=False,
                                                            binary_labels=v[0]['binary_labels'],
                                                            screwdriver_only=v[0]['screwdriver_only'])

            for i, d in enumerate(v):
                v[i]['test_x'] = test_x
                v[i]['test_y'] = test_y

            process_map(single_simulation, v, max_workers=n_cpu)

    elif run_mode == 'all':
        model_list = list_simulations(fill_mode=True, cycle_count=cycle_count)
        # model_list = model_list[:2]
        n_simulations = len(model_list)

        for i, d in enumerate(model_list):
            model_list[i]['cycle_count'] = cycle_count

        # Group models by the data settings - due to RAM constraints only models using identical datasets can be run
        # concurrently
        grouped_models = group_models(model_list)

        logger.info(f'Running {n_simulations} simulations in {len(grouped_models)} groups')

        for k, v in tqdm.tqdm(grouped_models.items(), desc='Simulation groups'):
            # Whether to reduce dimensionality
            if v[0]['n_dim'] < 125:
                reduce_dim = True
            else:
                reduce_dim = False

            _, _, test_x, test_y = aursad.get_dataset_numpy(path=Path(meta.data_path),
                                                            reduce_dimensionality=reduce_dim,
                                                            n_dimensions=n_dim,
                                                            subsample_data=True,
                                                            subsample_freq=2,
                                                            pad_data=False,
                                                            normal_samples=1,
                                                            damaged_samples=1,
                                                            assembly_samples=1,
                                                            missing_samples=1,
                                                            damaged_thread_samples=0,
                                                            loosening_samples=0,
                                                            drop_extra_columns=True,
                                                            onehot_labels=False,
                                                            binary_labels=binary_labels,
                                                            screwdriver_only=screwdriver_only)

            for i, d in enumerate(v):
                v[i]['test_x'] = test_x
                v[i]['test_y'] = test_y

            process_map(single_simulation, v, max_workers=n_cpu)


def single_simulation(simulation_args):
    """
    Run a single system simulation

    :param simulation_args:
    :return:
    """
    window = simulation_args['window']
    cycle_count = simulation_args['cycle_count']
    model_dir = simulation_args['model_dir']
    test_x = simulation_args['test_x']
    test_y = simulation_args['test_y']

    # The sensor domain
    domain = 0

    # The model's prediction horizon
    horizon = 1

    # The model's buffer size
    buffer = window

    # Create the device
    device = HapDev(buffer_size=buffer,
                    network_delay=meta.network_delay,
                    window=window,
                    horizon=horizon)

    # Create a sensor
    sensor = Sensor(domain=domain,
                    buffer_size=buffer,
                    dataset=test_x,
                    dataset_labels=test_y,
                    label_counter=window)
    device.add_sensor(sensor)

    # Load the ML model
    custom_objects = {'BL': BL,
                      'TABL': TABL,
                      'MaxNorm': tf.keras.constraints.max_norm}
    try:
        model = load_model(Path('../Zoo/Results/runs/' + model_dir + '/model'), custom_objects=custom_objects)
    except:
        model = load_model(Path(model_dir + '/model'), custom_objects=custom_objects)

    device.receive_model(model)

    # Lists to hold simulation results
    accuracy_list = []
    predicted_labels = []
    true_labels = []
    run_times = []
    debug_times = []

    # Run x steps of simulation
    # for i in tqdm.tqdm(range(cycle_count), desc='Running simulation cycles'):
    for i in range(cycle_count):
        if sensor.check_end():
            break
        accuracy, predicted_label, true_label, run_time, debug_results = device.run_one_cycle(domain)
        debug_times.append(debug_results)
        accuracy_list.append(accuracy)
        predicted_labels.append(predicted_label)
        true_labels.append(true_label)
        run_times.append(run_time)

    model_name = PurePath(model_dir)
    results_path = '../Results/' + model_name.name
    Path(results_path).mkdir(parents=False, exist_ok=True)

    # Save the simulation data
    simulation_results = {'True_labels': true_labels,
                          'Predicted_labels': predicted_labels,
                          'Accuracy': accuracy_list,
                          'Run_times': run_times}

    # Save debug data
    debug_df = pd.DataFrame(debug_times)
    debug_df.to_csv(results_path + '/debug_times.csv')

    simulation_results_df = pd.DataFrame(simulation_results)
    simulation_results_df.to_csv((results_path + '/Results_cycles-{cycle_count}_sensorID-{sensor}.csv').format(
            cycle_count=cycle_count,
            sensor=domain
    ), index_label='Cycle')

    # Plot the results
    plot_simulation_history(predicted_labels, true_labels, accuracy_list, run_times, results_path, domain, cycle_count)


if __name__ == '__main__':
    main()
